### PR TITLE
[Remove Stream/Optional usage] LPS-178143 portal-workflow-kaleo-service

### PR DESCRIPTION
**Pull request by JanisZhang**
Fix regression and resend, thanks!
previous:  https://github.com/liferay-core-infra/liferay-portal/pull/3827
https://issues.liferay.com/browse/LPS-178143

### CODE REVIEW

**Comment by JanisZhang** [comment]
ci:test:workflow

**Comment by zhangyan0701** [comment]
No regressions found.

**Comment by JanisZhang** [comment]
Thanks @zhangyan0701 !

File: modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/service/persistence/impl/KaleoTaskInstanceTokenFinderImpl.java [hunk 1]
```diff
@@ -916,15 +909,15 @@ protected void setTaskNames(
...
-		);
+		List<String> taskNameList = new ArrayList<>();
+
+		for (String taskName : taskNames) {
+			for (String keyword : _customSQL.keywords(taskName, false)) {
+				taskNameList.add(keyword);
+			}
+		}
+
+		taskNames = (String[])taskNameList.toArray();
```
Comment (dantewang) [inline modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/service/persistence/impl/KaleoTaskInstanceTokenFinderImpl.java:920]: @JanisZhang Please use `list.toArray(new String[0])`.

File: modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/service/persistence/impl/KaleoTaskInstanceTokenFinderImpl.java [hunk 2]
```diff
@@ -448,13 +447,10 @@ protected String getAssigneeClassPKs(
...
 
 		sb.append(
-			Stream.of(
-				assigneeClassPKs
-			).map(
-				String::valueOf
-			).collect(
-				Collectors.joining(StringPool.COMMA_AND_SPACE)
-			));
+			StringUtil.merge(
```
Comment (dantewang) [inline modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/service/persistence/impl/KaleoTaskInstanceTokenFinderImpl.java:450]: @JanisZhang The `StringBundler` usage is wrong before/after the change. The `StringBundler` is initialized with size of `array.length * 2 + 1`, but three's only 3 appends before calling `sb.toString()`.

For performance optimization, you should keep the initial size, use a loop to populate every class pk string form, and COMMA_AND_SPACE, reset the index to index - 1, and append the last element.

File: modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/service/impl/KaleoInstanceLocalServiceImpl.java [hunk 3]
```diff
@@ -393,18 +394,20 @@ kaleoDefinitionName, null, _getClassNames(assetClassName), null,
...
-			),
+			(List<KaleoInstance>)TransformUtil.transformToList(
+				hits.getDocs(),
+				document -> {
+					KaleoInstance kaleoInstance =
+						kaleoInstancePersistence.fetchByPrimaryKey(
+							GetterUtil.getLong(
+								document.get(Field.ENTRY_CLASS_PK)));
+
+					if (kaleoInstance == null) {
```
Comment (dantewang) [inline modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/service/impl/KaleoInstanceLocalServiceImpl.java:405]: @JanisZhang This is redundant, please simply. There are several places in this pull like this.

File: modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/internal/upgrade/v3_1_1/KaleoNotificationUpgradeProcess.java [hunk 4]
```diff
@@ -48,16 +48,19 @@ protected void doUpgrade() throws Exception {
...
+					StringUtil.split(resultSet.getString("notificationTypes")),
+					notificationType -> {
+						if (!Objects.equals(notificationType, "im") &&
+							!Objects.equals(
+								notificationType, "private-message")) {
+
+							return notificationType;
+						}
+
+						return null;
```
Comment (dantewang) [inline modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/internal/upgrade/v3_1_1/KaleoNotificationUpgradeProcess.java:61]: @JanisZhang Please return null in shortcut. By the way, please review your transformer usages, and put the lambda variable second in Objects.equals.

**Review by dantewang — CHANGES_REQUESTED** [review]
@JanisZhang Please resend, thanks!

**Pull request closed without merging**